### ISSUE TITLE
Excessive CPU usage & memory leak

### ISSUE THREAD

**Issue by kmpeters** [caqtdm/caqtdm#28]
Resizing the following ui file so that it it significantly smaller in both dimensions results in caQtDM using excessive CPU and leaking memory: [test.ui.txt](https://github.com/caqtdm/caqtdm/files/1304679/test.ui.txt).  If the ui file is returned to a larger size, the cpu usage returns to normal and the memory stops leaking. 

The problem requires at least two rows of buttons.  It doesn't occur if all of the caRelatedDisplays only have one item.

OS: RHEL7
caQtDM revision: bac201fc1c7634ccaad21301ae25c719db143cef
Qt version: 4.8.6
Qwt version: 6.1.3

This problem was discovered when a ui file with a caInclude widget included a ui file that was larger than the caInclude widget.  The included ui file was immediately resized which caused the problem to appear.

**Comment by AntonMezger** [comment]
I will take a look at this next week

Best regards

Anton

Envoyé de mon iPhone

Le 14 sept. 2017 à 23:53, Kevin Peterson <notifications@github.com<mailto:notifications@github.com>> a écrit :


Resizing the following ui file so that it it significantly smaller in both dimensions results in caQtDM using excessive CPU and leaking memory: test.ui.txt<https://github.com/caqtdm/caqtdm/files/1304679/test.ui.txt>. If the ui file is returned to a larger size, the cpu usage returns to normal and the memory stops leaking.

The problem requires at least two rows of buttons. It doesn't occur if all of the caRelatedDisplays only have one item.

OS: RHEL7
caQtDM revision: bac201f<https://github.com/caqtdm/caqtdm/commit/bac201fc1c7634ccaad21301ae25c719db143cef>
Qt version: 4.8.6
Qwt version: 6.1.3

**Comment by HelgeBrands** [comment]
on Windows with QT 5.8 and 4.8.5 it is not reproduceable

**Comment by kmpeters** [comment]
I forgot to mention that I can't reproduce the problem with older versions of caQtDM (v3.8.10 and v4.0.0).

**Comment by AntonMezger** [comment]
Dear Kevin

Try to add the following statements in the construction of caRowColMenu:

If(minimumHeight() < 1) setMinimumHeight(20);
If(minimumWidth() < 1) setMinimumWidth(40);

In my case that solved the problem of an avalanche of resizeevents. Please try it in your case if it solves your problem.

Best regards

Anton

**Comment by kmpeters** [comment]
Adding those lines makes it harder to reproduce the problem, but I'm still able to get caQtDM into an excessive-CPU state.  Without those lines caQtDM would use excessive CPU the first time the screen was resized to be significantly smaller.  With those lines I have to resize the screen a few times (smaller -> bigger -> smaller) before the problem occurs.

**Comment by AntonMezger** [comment]
Dear Kevin,

The problem has been solved. You may download imagepushbutton in order to solve the problem.

Best regards

Anton

**Comment by kmpeters** [comment]
There is still a problem.  I can put ef55f72e1ea0e5f1066bd2734b5f51af505060b7 into a the excessive-cpu state with a single resize  (the buttons need to be resized to be very small, but still visible).  A gdb backtrace during the excessive cpu usage is attached: [Excessive-CPU-caQtDM-debug.txt](https://github.com/caqtdm/caqtdm/files/1318938/Excessive-CPU-caQtDM-debug.txt)

**Comment by AntonMezger** [comment]
I could not reproduce the excessive cpu load any more with my last change. I will take one more look.

Sure, you rebuild everything?

Best regards

Anton

Envoyé de mon iPhone

Le 20 sept. 2017 à 21:15, Kevin Peterson <notifications@github.com<mailto:notifications@github.com>> a écrit :


There is still a problem. I can put ef55f72<https://github.com/caqtdm/caqtdm/commit/ef55f72e1ea0e5f1066bd2734b5f51af505060b7> into a the excessive-cpu state with a single resize (the buttons need to be resized to be very small, but still visible). A gdb backtrace during the excessive cpu usage is attached: Excessive-CPU-caQtDM-debug.txt<https://github.com/caqtdm/caqtdm/files/1318938/Excessive-CPU-caQtDM-debug.txt>

**Comment by kmpeters** [comment]
Everything has been rebuilt.  I started fresh on a different linux machine.  I built Qt 4.8.7 and Qwt 6.1.3 locally, with debugging on (otherwise the gdb backtrace isn't helpful).  I include the output of the ldd command in my script that starts gdb so I can confirm I'm using the correct shared libraries.

This is the ui file before the resize event: ![before](https://user-images.githubusercontent.com/10675924/30670112-b191044c-9e26-11e7-838f-d2d2c845ebc7.png)
This is the ui file after the resize event: ![after](https://user-images.githubusercontent.com/10675924/30670115-b316cfcc-9e26-11e7-8376-22f4819e0176.png)

It only takes the single resize event for the CPU usage to spike and the memory to begin leaking.

**Comment by AntonMezger** [comment]
I made a dirty correction in order to prevent firing of resize events for one pixel difference in size. I think that solves the problem, but probably it should be done in another way

Best regards

Envoyé de mon iPhone

Le 21 sept. 2017 à 00:13, Kevin Peterson <notifications@github.com<mailto:notifications@github.com>> a écrit :


Everything has been rebuilt. I started fresh on a different linux machine. I built Qt 4.8.7 and Qwt 6.1.3 locally, with debugging on (otherwise the gdb backtrace isn't helpful). I include the output of the ldd command in my script that starts gdb so I can confirm I'm using the correct shared libraries.

This is the ui file before the resize event: [before] <https://user-images.githubusercontent.com/10675924/30670112-b191044c-9e26-11e7-838f-d2d2c845ebc7.png>
This is the ui file after the resize event: [after] <https://user-images.githubusercontent.com/10675924/30670115-b316cfcc-9e26-11e7-8376-22f4819e0176.png>

It only takes the single resize event for the CPU usage to spike and the memory to begin leaking.

**Comment by kmpeters** [comment]
The problem still exists when I use commit d1a7c253c886faf539f7e459e61f47826993a08c.

I slowly reduced the size of test.ui in both dimensions, stopping frequently to check the cpu usage of caQtDM.  caQtDM used a normal amount of cpu most of the times that I stopped.  If I noticed the font size of the lower button shrink significantly when the font of the top button stayed the same, I always saw excessive cpu usage.  Here is a screenshot illustrating the font-size difference:
![fontdifference](https://user-images.githubusercontent.com/10675924/30765785-feb61bc2-9fb6-11e7-9375-c9bdef3156dc.png)

I've been using the debug builds of Qt and Qwt during my testing.  I'll build non-debug versions next week to see if the problem goes away with optimized libraries.

If I build a newer version of Qt (> 4.8), which version should I use?

**Comment by AntonMezger** [comment]
I will look into it on wednesday



Envoyé de mon iPhone

Le 22 sept. 2017 à 23:58, Kevin Peterson <notifications@github.com<mailto:notifications@github.com>> a écrit :


The problem still exists when I use commit d1a7c25<https://github.com/caqtdm/caqtdm/commit/d1a7c253c886faf539f7e459e61f47826993a08c>.

I slowly reduced the size of test.ui in both dimensions, stopping frequently to check the cpu usage of caQtDM. caQtDM used a normal amount of cpu most of the times that I stopped. If I noticed the font size of the lower button shrink significantly when the font of the top button stayed the same, I always saw excessive cpu usage. Here is a screenshot illustrating the font-size difference:
[fontdifference]<https://user-images.githubusercontent.com/10675924/30765785-feb61bc2-9fb6-11e7-9375-c9bdef3156dc.png>

I've been using the debug builds of Qt and Qwt during my testing. I'll build non-debug versions next week to see if the problem goes away with optimized libraries.

If I build a newer version of Qt (> 4.8), which version should I use?
